### PR TITLE
CX: Handle goal executability via pddl preconditions

### PR DESCRIPTION
**Pull request by TarikViehmann**
This PR mainly aims to make a cleaner executability check via pddl preconditions.
In order to describe the required conditions in strips-style pddl, it introduces several auxiliary domain predicates that hide a way most of the domain complexity.
These are very similar to derived predicates in pddl, although lacking an actual description in pddl as there is no automatic translation for that yet.

### PDDL Goal Preconditions
 
The core idea for the grounding is to:
1) introduce domain operators of the style `goal-<goal-class>` where the signature _almost_ matches the goal params of the respective goals from the class `<goal-class>`. _Almost_, as we currently model the assignment of robots to goals via the goal meta instead of goal params. As I did not want to touch this fundamental piece, the first param of each goal operator is `r`, which takes a robot that is grounded from the respective `asigned-to` field of the goal-meta.
2) Ground and "unground" the goals whenever their signature (`params` and `assigned-to` fields) changes.
3) Replace existing goal executability checks by the automatic pddl formula checks.

### Changes to existing goals

In order to implement this, I cleaned up the existing parameters of goals to get rid of hacky solutions (such as `UNKNOWN` workpieces or `wp-loc` and `wp-side` params that were simply false because the workpiece in question was already being held).
The main changes regarding this are:
1) no more `wp-loc` `wp-side` and `target-side` params.
2) no more referencing of unnamed cap-carriers, instead, hide this behind indirect referencing if needed (e.g., "_any cap carrier with this cap color_") and resort to explicit referencing if appropriate (e.g., instead of formulating `DISCARD` and `PAY-FOR-RINGS-WITH-CAP-CARRIER` goals before a cap-less carrier is cleaned, simply formulate them once they are blocking the CS output. Then `PAY-FOR-RINGS-WITH-CAP-CARRIER` actually is equivalent to `PAY-FOR-RINGS-WITH-BASE`, hence we only need one kind of payment goal (assuming we do not want to pay using cap-carriers from shelfs).

Further, some of the existing behavior is not ported/altered through these changes, afaik these are the differences:
1) slide payment and mount ring goals now should work in parallel and multiple slide payment goals can target the same machine at once. For the duration of a running instruct goal on an RS, `wp-put-slide-cc` actions are not executable, similarly, in order to start a `prepare-rs` action, no `wp-put-slide-cc` can run. However, I explicitly prevent the selection of 3 payment goals to the same machine, primarily because encoding this properly is ugly; but also because this is likely never a good idea anyways.
2) There is no explicit deadlock prevention when handing ring stations for a scenario, where both outputs are full and the workpieces on the outputs need the opposing RS, then there should always be one RS input that is kept free. Encoding this through the new derived predicates would be very volatile, as it requires precise reasoning over multiple formulated and running goals. As this deadlock can only appear when at least 4 products use ring stations at the same time, this will not happen unless we try to build at least 4 C1-C3 products in parallel, which we do not want to do anyways.
3) There currently is no explicit mechanism in the goal preconditions  that ensure that mount ring goals for a workpiece are started in-order. It is handled later, during the aggregation of executable goals (as they are siblings in a `run-all` sub-tree). But I think it would be worth considering a more explicit way of expressing the order in the preconditions as well.
4) The `goal-in-retry-period` mechanism is currently not implemented. It was not flashed out so far, anyways. Since I anticipate an overhaul of these kind of execution monitoring mechanisms, I did not bother to port this.

### Updated dynamic goal priorities and strategy

The prioritization of tasks needs to be adjusted, but will be a separate PR.

~As the above changes messed with the existing goal priorities, some updates were implemented:~
1) ~Extend dynamic priority increases from just clearing CS outputs to also apply to buffer cap and payment goals.~
2) ~Properly use the highest possible priority increase if multiple ongoing orders require the step (e.g., in case of ring payments)~
3) ~Retroactively apply priority increases to goals that are formulated after priority increase is asserted~

~While at it, I also added a small improvement to the overall order selection strategy, as the workload calculations did not consider steps of a product that were already finished.~

### Misc
I bumped the saliences of rules that handle incoming refbox data to make sure that our domain is always as up-to-date as possible.
We should probably also do the same for wp meta and order meta calculations at some point in the future.

Also, I made some slight improvements to the goal restriction mechanism that aims at keeping robots committed to one product assembly sequence. However, we likely need to tweak it more, as this mainly prevents other robots from grabbing the respective goals, but does not prevent the restricted robot to do other tasks. This could actually lead to slow-downs if said robot decides to do something else while keeping the control of the assembly sequence.

This PR needs to update the submodule pointer once https://github.com/fawkesrobotics/fawkes/pull/381 is merged.

### CODE REVIEW

File: src/clips-specs/rcll-central/fixed-sequence.clp [hunk 1]
```diff
@@ -335,231 +302,261 @@
...
+				(plan-assert-action wp-put-slide-cc ?robot
+				 ?wp ?target-mps ?rs-before ?rs-after)
 			)
 		)
 	)
 	(modify ?g (mode EXPANDED))
 )
 
-(defrule goal-expander-get-shelf-to-fill-rs
+; (defrule goal-expander-get-cap-carrier-to-fill-rs-holding
```
Comment (snoato) [inline src/clips-specs/rcll-central/fixed-sequence.clp:468]: we should probably remove these blocks

File: src/clips-specs/rcll-central/goal-production.clp [hunk 2]
```diff
@@ -541,7 +528,7 @@
   (bind ?goal
     (goal-tree-assert-central-run-all-prio PRODUCE-ORDER ?*PRODUCTION-C0-PRIORITY*
       (goal-production-assert-deliver ?wp-for-order ?order-id ?instruct-parent ?ds)
-      (goal-production-assert-mount-cap ?wp-for-order ?cs ?bs INPUT ?order-id)
+      (goal-production-assert-mount-cap ?wp-for-order ?cs ?col-cap ?order-id)
```
Comment (snoato) [inline src/clips-specs/rcll-central/goal-production.clp:531]: maybe we should change it to cap-col for consistency with other parts of the code?

Comment (TarikViehmann) [reply src/clips-specs/rcll-central/goal-production.clp]: fixed it throughout the code

File: src/clips-specs/rcll-central/fixed-sequence.clp (hunk 1 above)
Comment (TarikViehmann) [reply src/clips-specs/rcll-central/fixed-sequence.clp]: true, good catch

**Review by snoato — APPROVED** [review]

**Comment by TarikViehmann** [comment]
Thank you very much!

**Pull request merged**